### PR TITLE
Issue #1071: Fix race condition while creating directory

### PR DESCRIPTION
**Pull request by Schrank**

### CODE REVIEW

File: src/DataPool/UrlKeyStore/FileUrlKeyStore.php [hunk 1]
```diff
@@ -103,8 +104,10 @@ private function formatRecordToWrite(string $urlKey, string $contextData, string
 
     private function ensureDirectoryExists(string $directoryPath)
     {
-        if (!file_exists($directoryPath)) {
-            mkdir($directoryPath, 0700, true);
+        if (! @mkdir($directoryPath, 0700, true) && ! is_dir($directoryPath)) {
+            throw new DirectoryDoesNotExistException(
```
Comment (tim-bezhashvyly) [inline src/DataPool/UrlKeyStore/FileUrlKeyStore.php]: I think using this exception class is confusing.

File: src/DataPool/UrlKeyStore/FileUrlKeyStore.php [hunk 2]
```diff
@@ -103,8 +104,10 @@ private function formatRecordToWrite(string $urlKey, string $contextData, string
...
+                sprintf('Directory "%s" was not found and could not be created.', $directoryPath)
```
Comment (tim-bezhashvyly) [inline src/DataPool/UrlKeyStore/FileUrlKeyStore.php]: Exception message also not quite informative. Reads like "Mein Name ist Hase .." to me.

File: tests/Unit/Suites/DataPool/UrlKeyStore/FileUrlKeyStoreTest.php [hunk 3]
```diff
@@ -37,9 +43,22 @@ private function prepareTemporaryStorage() : string
         return $temporaryStoragePath;
     }
 
+    protected function setUp()
```
Comment (tim-bezhashvyly) [inline tests/Unit/Suites/DataPool/UrlKeyStore/FileUrlKeyStoreTest.php]: Please make final.

File: tests/Unit/Suites/DataPool/UrlKeyStore/FileUrlKeyStoreTest.php [hunk 4]
```diff
@@ -37,9 +43,22 @@ private function prepareTemporaryStorage() : string
...
+
     protected function tearDown()
     {
-        (new LocalFilesystem())->removeDirectoryAndItsContent($this->temporaryStoragePath);
+        try {
+            (new LocalFilesystem())->removeDirectoryAndItsContent($this->temporaryStoragePath);
+        } catch (DirectoryDoesNotExistException $e) {
+            if (self::$createDirectory) {
+                throw $e;
+            }
```
Comment (tim-bezhashvyly) [inline tests/Unit/Suites/DataPool/UrlKeyStore/FileUrlKeyStoreTest.php]: I'm not sure what are you trying to achieve here.

Comment (Schrank) [inline tests/Unit/Suites/DataPool/UrlKeyStore/FileUrlKeyStoreTest.php]: `-        (new LocalFilesystem())->removeDirectoryAndItsContent($this->temporaryStoragePath);
` throws an exception if the directory doens't exist. `self::$createDirectory` tells us, wether we expect the directory. Will add a method to make it clearer.

File: src/DataPool/UrlKeyStore/FileUrlKeyStore.php (hunk 1 above)
Comment (Schrank) [inline src/DataPool/UrlKeyStore/FileUrlKeyStore.php]: What do you think is better? It is a FileBasedStorage - a FilebaseDirectoryNotFoundException is intuitive imho :-) But you are right, I'll fix the exception message

**Comment by Schrank** [comment]
Please have a second look @tim-bezhashvyly

File: src/DataPool/UrlKeyStore/FileUrlKeyStore.php (hunk 1 above)
Comment (tim-bezhashvyly) [inline src/DataPool/UrlKeyStore/FileUrlKeyStore.php]: Maybe `CanNotCreateDirectoryException` or is it something bigger?

File: src/DataPool/UrlKeyStore/FileUrlKeyStore.php [hunk 5]
```diff
@@ -103,8 +104,14 @@ private function formatRecordToWrite(string $urlKey, string $contextData, string
 
     private function ensureDirectoryExists(string $directoryPath)
     {
-        if (!file_exists($directoryPath)) {
-            mkdir($directoryPath, 0700, true);
+        if (! @mkdir($directoryPath, 0700, true) && ! is_dir($directoryPath)) {
+            throw new DirectoryDoesNotExistException(
+                sprintf(
```
Comment (tim-bezhashvyly) [inline src/DataPool/UrlKeyStore/FileUrlKeyStore.php]: You could in fact place `sprintf` at the same line with the exception class. Would be 2 lines shorter. 😜

File: tests/Unit/Suites/DataPool/UrlKeyStore/FileUrlKeyStoreTest.php [hunk 6]
```diff
@@ -14,18 +15,28 @@
...
+
+    /**
      * @var string
      */
     private $temporaryStoragePath;
 
-    final protected function createUrlKeyStoreInstance() : FileUrlKeyStore
+    private static function isDirectoryExpected()
+    {
+        return self::$createDirectory;
```
Comment (tim-bezhashvyly) [inline tests/Unit/Suites/DataPool/UrlKeyStore/FileUrlKeyStoreTest.php]: Maybe something like `$isDirectoryCreated`?

Comment (Schrank) [inline tests/Unit/Suites/DataPool/UrlKeyStore/FileUrlKeyStoreTest.php]: Hard to say. When I created the property and still, the property decides wether the directory will be created. To know wether the directory will be created there is a method.

So I think the name is good. But I'm fine changing it.

**Pull request merged**